content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown

```diff
@@ -0,0 +1,380 @@
+can just set it back to 0 later.
+
+In our case, we had debug info and a debugger, so finding where the symbol was
+located in memory is pretty easy.  There's also `/proc/kallsyms` which is great
+if it's enabled.  Linux on Power doesn't yet support KASLR which also saves us a
+headache or two here, and you can [feel free to ask me why it's low on the
+priority list](https://forums.grsecurity.net/viewtopic.php?f=7&t=3367).
+
+So now we have a null pointer dereference.  Now let's get that doing something!
+
+## Stage 4: preparing the exploit
+
+Virtual memory is one heck of a drug.  If the kernel is going to execute from
+`0x0`, we just need to `mmap()` to 0!  Easy.
+
+Well, it's not that easy.  Turning any null pointer dereference into an easy
+attack vector is not ideal, so users aren't allowed to mmap to low address
+ranges, in our case, below `PAGE_SIZE`.  Surely there's nothing in the kernel
+that would try to dereference a pointer + `PAGE_SIZE`?  Maybe that's for a
+future CTF...
+
+There's a sysctl for this, so in the actual CTF we just did `sysctl -w
+vm.mmap_min_addr=0` and moved on for brevity.  As I was writing this I decided
+to make sure it was possible to bypass this without cheating by making use of
+our kernel write primitive, and sure enough, it works!  I had to zero out both
+`mmap_min_addr` and `dac_mmap_min_addr` symbols, the latter seemingly required
+for filesystem interactions to work post-exploit.
+
+So now we can trigger a null pointer dereference in the kernel and we can
+`mmap()` our shellcode to 0x0, we should probably get some shellcode.  We want
+to escalate our privileges, and the easiest way to do that is the iconic
+`commit_creds(prepare_kernel_cred(0))`.  As of somewhat recently it's a bit more
+complex than that, but we're still back in v4.13, so we just need a way to
+execute that from a triggered null pointer dereference.
+
+## Stage 5: the shellcode
+
+The blessing and curse of Power being a niche architecture is that it's hard to
+find existing exploits for.  Perhaps lacking in grace and finesse, but effective
+nonetheless, is the shellcode I wrote myself:
+
+```c
+	static const unsigned char shellcode[] = {
+		0x00, 0x00, 0xc0, 0x3b, // li r30, 0
+		0x20, 0x00, 0x9e, 0xe9, // ld r12,32(r30)
+		0x00, 0x00, 0xcc, 0xfb, // std r30,0(r12)
+		0x18, 0x00, 0x9e, 0xe9, // ld r12,24(r30)
+		0xa6, 0x03, 0x89, 0x7d, // mtctr r12
+		0x20, 0x04, 0x80, 0x4e, // bctr
+	};
+```
+
+After the CTF I encouraged everyone to try writing their own shellcode and noone
+did, and I will take that as a sign that mine is flawlessly designed.
+
+First we throw 0 into `r30`, which sounds like a register we'll get away with
+clobbering.  We load an offset of 32 bytes from the value of `r30` into `r12`
+(and `r30` is 0, so this is the address 32).  Then, we store the value of `r30`
+(0) into the address in `r12`.
+
+(we write zero to the address placed at 32 / 0x20)
```

Comment:
this looks odd?

Reply:
it does but I felt like the explanation was overcomplicated so I wanted to summarise it, I can try and do that in a different way but I'm not exactly sure how